odoo/addons/base/transpiler/sourcemap_generator.py

```diff
@@ -0,0 +1,191 @@
+        }
+        if (self._file is not None):
+            map["file"] = self._file
+
+        if (self._sourceRoot is not None):
+            map["sourceRoot"] = self._sourceRoot
+
+        return map
+
+    def add_source(self, source_name, source_content, previous_genererated_line, nb_skip_line_output=0):
+        """Adds a new source file in the sourcemap. All the lines of the source file will be mapped line by line
+        to the generated file from the (previous_genererated_line + nb_skip_line_output). All lines between
+        previous_genererated_line and (previous_genererated_line + nb_skip_line_output) will
+        be mapped to line 1 of the source file.
+
+        Example:
+            ls 1 = Line 1 from new source file
+            lg 1 = Line 1 from genereted file
+            ls 1 <=> lg 1 Line 1 from new source file is map to  Line 1 from genereted file
+            nb_ls = number of lines in the new source file
+
+            Step 1:
+            ls 1 <=> lg previous_genererated_line + 1
+
+            Step 2:
+            ls 1 <=> lg previous_genererated_line + nb_skip_line_output + 1
+            ls 2 <=> lg previous_genererated_line + nb_skip_line_output + 2
+            ...
+            ls nb_ls <=> lg previous_genererated_line + nb_skip_line_output + nb_ls
+
+
+        :param source_name: name of the source to add
+        :param source_content: content of the source to add
+        :param previous_genererated_line: Line where we start to map the new source
+        :param nb_skip_line_output: Number of lines to pass in the generated file before starting mapping line by line
+
+        :return number of lines added
+        """
+        source_line_count = len(source_content.split("\n"))
+
+        if (source_name not in self._sources):
+            self._sources[source_name] = len(self._sources)
+
+        self._sourcesContents[source_name] = source_content
+        if nb_skip_line_output > 0:
+            # adds a mapping between the first line of the source
+            # and the first line of the corresponding code in the generated file.
+            self._mappings.append({
+                "generatedLine": previous_genererated_line + 1,
+                "originalLine": 1,
+                "source": source_name,
+            })
+        for i in range(1, source_line_count + 1):
+            self._mappings.append({
+                "generatedLine": previous_genererated_line + i + nb_skip_line_output,
+                "originalLine": i,
+                "source": source_name,
+            })
+        return source_line_count + nb_skip_line_output
+
+
+def memorize(fn):
+    """
+    Provide memoization.
+    It is used to avoid useless computation of the mapping.
+    """
+    cache = {}
+    def fn_memoize(args):
+        if args not in cache:
+            cache[args] = fn(args)
+        return cache[args]
+    return fn_memoize
+
+
+def base64vlq():
```

Comment:
Wouldn't it be possible/simpler to reuse the standard module `base64`?

Reply:
I think base64 does not have a base64VLQ implementation. I'm not an expert but there is a difference between the two

Reply:
@rco-odoo yes I was wondering the same but it looks like the only base64 part of base64vlq is its use of the base64 alphabet: base64vlq encodes integers (or sequence thereof but that seems to encode integers individually), not bytes sequences, and the results have nothing in common e.g. `123` is `'2H'` in base64VLQ. In base64, `bytes([123])` is `'ew=='` and `b'123'` is `'MTIz'`.